lib/src/embedded/compiler.ts

```diff
@@ -118,10 +84,132 @@ export class EmbeddedCompiler {
       .subscribe(() => process.nextTick(() => this.close()));
   }
 
+  /**
+   * Dispatches a compile request. Returns a promise that resolves when a
+   * compile response with a matching ID is received from the embedded compiler.
+   */
+  compile(request: InboundMessage.CompileRequest) {
+    const id = this.nextRequestId();
+    request.setId(id);
+    const message = new InboundMessage();
+    message.setCompilerequest(request);
+
+    this.recordRequest(id);
+    this.messageTransformer.write$.next(message);
+
+    return this.messageTransformer.read$
+      .pipe(
+        filter(message => {
+          return (
+            message.hasCompileresponse() &&
+            message.getCompileresponse()!.getId() === id
+          );
+        }),
+        map(message => message.getCompileresponse()!),
+        tap(response => this.recordResponse(response.getId())),
+        take(1),
+        takeUntil(this.error$)
+      )
+      .toPromise();
+  }
+
+  listen(
+    messageType: OutboundMessage.MessageCase,
+    processRequest: (request: Request) => Response
+  ) {
+    this.messageTransformer.read$
+      .pipe(
+        filter(message => message.getMessageCase() === messageType),
+        map(message => getRequest(message)),
+        tap(request => this.recordRequest(request.getId())),
```

Comment:
Do you need to record the ID here? You're allowed to have your `InboundRequest` IDs overlap with `OutboundRequest` IDs.

Reply:
I'm recording it here in case the compiler sends more than one request with with the same request ID. Is that something we want to protect against?

Reply:
I think that's reasonable to check, but I'd probably use a different set for inbound IDs and outbound IDs—otherwise, a reader (like me!) is liable to get confused about why you're tracking outbound IDs at all.